Applications/Piano/AudioEngine.cpp

```diff
@@ -0,0 +1,217 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "AudioEngine.h"
+#include <limits>
+#include <math.h>
+
+AudioEngine::AudioEngine()
+{
+}
+
+AudioEngine::~AudioEngine()
+{
+}
+
+void AudioEngine::fill_buffer(Sample* buffer)
+{
+    memset(buffer, 0, buffer_size);
+
+    for (int i = 0; i < sample_count; ++i) {
+        for (size_t note = 0; note < note_count; ++note) {
+            if (!m_note_on[note])
+                continue;
+            double val = 0;
+            switch (m_wave) {
+            case Wave::Sine:
+                val = ((volume * m_power[note]) * sine(note));
+                break;
+            case Wave::Saw:
+                val = ((volume * m_power[note]) * saw(note));
+                break;
+            case Wave::Square:
+                val = ((volume * m_power[note]) * square(note));
+                break;
+            case Wave::Triangle:
+                val = ((volume * m_power[note]) * triangle(note));
+                break;
+            case Wave::Noise:
+                val = ((volume * m_power[note]) * noise());
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+            }
+            buffer[i].left += val;
+        }
+        buffer[i].right = buffer[i].left;
+    }
+
+    if (m_decay) {
+        for (size_t note = 0; note < note_count; ++note) {
+            if (m_note_on[note]) {
+                m_power[note] -= m_decay / 100.0;
+                if (m_power[note] < 0)
+                    m_power[note] = 0;
+            }
+        }
+    }
+
+    if (m_delay) {
+        if (m_delay_buffers.size() >= m_delay) {
+            auto to_blend = m_delay_buffers.dequeue();
+            for (int i = 0; i < sample_count; ++i) {
+                buffer[i].left += (*to_blend)[i].left * 0.333333;
+                buffer[i].right += (*to_blend)[i].right * 0.333333;
+            }
+        }
+
+        auto delay_buffer = make<FixedArray<Sample>>(sample_count);
+        for (int i = 0; i < sample_count; ++i)
+            (*delay_buffer)[i] = buffer[i];
```

Comment:
Can we use `memcpy()` here instead?

Reply:
`FixedArray` doesn't have anything like `data()`
We could add that?
...Or we can memcpy to the address of [0]